### ISSUE TITLE
Issues emulating nc utility in armv7 environment

### ISSUE THREAD

**Issue by GarretReece** [trailofbits/manticore#587]
### Manticore version
git commit fccf451 on the master branch

### Summary of the problem
Incorrect instruction semantics for Thumb mode

### Step to reproduce the behavior
`$ manticore -- ./nc-android-ndk-armv7-debug -p 1234`

### Expected behavior
Thumb mode instructions supported

### Actual behavior
* incorrect SUB semantics for thumb mode
* incorrect mode exchanges due to BLX instructions
* gettid syscall not implemented

### Any relevant logs
Running with the latest in master
```
$ manticore -- ./nc-android-ndk-armv7-debug -p 1234 
2017-11-22 07:46:27,353: [125542] m.manticore:INFO: Loading program ./nc-android-ndk-armv7-debug
2017-11-22 07:46:28,645: [125542] m.manticore:INFO: Starting 1 processes.
2017-11-22 07:46:28,716: [125542] m.c.executor:ERROR: Exception: SUB() takes exactly 4 arguments (3 given)
Traceback (most recent call last):
  File "/home/trailofbits/workspace/manticore/manticore/core/executor.py", line 456, in run
    if not current_state.execute():
  File "/home/trailofbits/workspace/manticore/manticore/core/state.py", line 126, in execute
    result = self._platform.execute()
  File "/home/trailofbits/workspace/manticore/manticore/platforms/linux.py", line 2058, in execute
    self.current.execute()
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/abstractcpu.py", line 759, in execute
    getattr(self, name)(*insn.operands)
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/abstractcpu.py", line 849, in new_method
    return old_method(cpu, *args, **kw_args)
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/arm.py", line 51, in instruction_implementation
    ret = body(cpu, *args, **kwargs)
TypeError: SUB() takes exactly 4 arguments (3 given)

Exception: SUB() takes exactly 4 arguments (3 given)
Traceback (most recent call last):
  File "/home/trailofbits/workspace/manticore/manticore/core/executor.py", line 456, in run
    if not current_state.execute():
  File "/home/trailofbits/workspace/manticore/manticore/core/state.py", line 126, in execute
    result = self._platform.execute()
  File "/home/trailofbits/workspace/manticore/manticore/platforms/linux.py", line 2058, in execute
    self.current.execute()
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/abstractcpu.py", line 759, in execute
    getattr(self, name)(*insn.operands)
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/abstractcpu.py", line 849, in new_method
    return old_method(cpu, *args, **kw_args)
  File "/home/trailofbits/workspace/manticore/manticore/core/cpu/arm.py", line 51, in instruction_implementation
    ret = body(cpu, *args, **kwargs)
TypeError: SUB() takes exactly 4 arguments (3 given)

2017-11-22 07:46:28,719: [125542] m.manticore:INFO: Results in /home/trailofbits/workspace/nc/mcore_Lso07U
2017-11-22 07:46:28,719: [125542] m.manticore:INFO: Total time: 0.0734069347382
```

[nc-android-ndk-armv7-debug.tar.gz](https://github.com/trailofbits/manticore/files/1496569/nc-android-ndk-armv7-debug.tar.gz)

**Comment by mossberg** [comment]
@yan, when you're back, can you help with review for this?

**Comment by GarretReece** [comment]
Fixed via PR #588.
